test/Dialect/FIRRTL/infer-domains.mlir

```diff
@@ -0,0 +1,323 @@
+  ) {
+    %c = firrtl.wire : !firrtl.uint<1>
+
+    // %c should infer domain sequence [%A, %B]
+    firrtl.matchingconnect %c, %a : !firrtl.uint<1>
+
+    // This should be legal since %c has inferred [%A, %B]
+    firrtl.matchingconnect %d, %c : !firrtl.uint<1>
+  }
+}
+
+// Domain duplicate equivalence - should be legal.
+// CHECK-LABEL: DuplicateDomainEquivalence
+firrtl.circuit "DuplicateDomainEquivalence" {
+  firrtl.domain @ClockDomain
+  firrtl.module @DuplicateDomainEquivalence(
+    in %A: !firrtl.domain of @ClockDomain,
+    in %a: !firrtl.uint<1> domains [%A, %A],
+    out %b: !firrtl.uint<1> domains [%A]
+  ) {
+    // This should be legal since duplicate domains are canonicalized.
+    firrtl.matchingconnect %b, %a : !firrtl.uint<1>
+  }
+}
+
+// Unsafe domain cast with sequences
+// CHECK-LABEL: UnsafeSequenceCast
+firrtl.circuit "UnsafeSequenceCast" {
+  firrtl.domain @ClockDomain
+  firrtl.domain @PowerDomain
+
+  firrtl.module @UnsafeSequenceCast(
+    in %C1: !firrtl.domain of @ClockDomain,
+    in %C2: !firrtl.domain of @ClockDomain,
+    in %P1: !firrtl.domain of @PowerDomain,
+    in  %i: !firrtl.uint<1> domains [%C1, %P1],
+    out %o: !firrtl.uint<1> domains [%C2, %P1]
+  ) {
+    %0 = firrtl.unsafe_domain_cast %i domains %C2 : !firrtl.uint<1>
+    firrtl.matchingconnect %o, %0 : !firrtl.uint<1>
+  }
+}
+
+//  Different port types domain inference.
+// CHECK-LABEL: DifferentPortTypes
+firrtl.circuit "DifferentPortTypes" {
+  firrtl.domain @ClockDomain
+  firrtl.module @DifferentPortTypes(
+    in %A: !firrtl.domain of @ClockDomain,
+    in %uint_input: !firrtl.uint<8> domains [%A],
+    in %sint_input: !firrtl.sint<4> domains [%A],
+    out %uint_output: !firrtl.uint<8>,
+    out %sint_output: !firrtl.sint<4>
+  ) {
+    firrtl.matchingconnect %uint_output, %uint_input : !firrtl.uint<8>
+    firrtl.matchingconnect %sint_output, %sint_input : !firrtl.sint<4>
+  }
+}
+
+// Domain inference through wires.
+// CHECK-LABEL: DomainInferenceThroughWires
+firrtl.circuit "DomainInferenceThroughWires" {
+  firrtl.domain @ClockDomain
+  firrtl.module @DomainInferenceThroughWires(
+    in %A: !firrtl.domain of @ClockDomain,
+    in %input: !firrtl.uint<1> domains [%A],
+    // CHECK: out %output: !firrtl.uint<1> domains [%A]
+    out %output: !firrtl.uint<1>
+  ) {
+    %wire1 = firrtl.wire : !firrtl.uint<1>
+    %wire2 = firrtl.wire : !firrtl.uint<1>
+
+    firrtl.matchingconnect %wire1, %input : !firrtl.uint<1>
+    firrtl.matchingconnect %wire2, %wire1 : !firrtl.uint<1>
+    firrtl.matchingconnect %output, %wire2 : !firrtl.uint<1>
+  }
+}
+
+// Export: add output domain port for domain created internally.
+// CHECK-LABEL: ExportDomain
+firrtl.circuit "ExportDomain" {
+  firrtl.domain @ClockDomain
+
+  firrtl.extmodule @Foo(
+    out A: !firrtl.domain of @ClockDomain,
+    out o: !firrtl.uint<1> domains [A]
+  )
+
+  firrtl.module @ExportDomain(
+    // CHECK: out %ClockDomain: !firrtl.domain of @ClockDomain
```

Comment:
Getting a better name for this may be reasonable. Just "ClockDomain" is going to run into `ClockDomain_0`, etc. (Though, how many different inferred domains like this are going to exist?)

Possible something like `domain_foo_A` to give a hint as to its provenance.